Lazy progress + Progress Groups

sipsma opened the pull request:
Example output for the following LLB (built w/ a local export in order to force unlazying):
```
pg := llb.ProgressGroup("group", "TestGroup")
base := llb.Image("busybox")
src := llb.Image("alpine").
		File(llb.Mkfile("/foo", 0755, nil)).
		File(llb.Mkfile("/bar", 0755, nil))
a := llb.Scratch().File(llb.Copy(src, "/foo", "/a"), pg)
b := llb.Scratch().File(llb.Copy(src, "/bar", "/b"), pg)
m := llb.Merge([]llb.State{base, a, b}, pg)
dt, err := m.Marshal(context.TODO())
```

Here's the output when using `tty` mode:
```
[+] Building 0.8s (9/9) FINISHED
 => docker-image://docker.io/library/alpine:latest                                                                                                                                      0.4s
 => => resolve docker.io/library/alpine:latest                                                                                                                                          1.4s
 => => sha256:9b3977197b4f2147bdd31e1271f811319dcd5c2fc595f14e81f5351ab6275b99 2.72MB / 2.72MB                                                                                          0.4s
 => => extracting sha256:9b3977197b4f2147bdd31e1271f811319dcd5c2fc595f14e81f5351ab6275b99                                                                                               0.1s
 => docker-image://docker.io/library/busybox:latest                                                                                                                                     0.2s
 => => resolve docker.io/library/busybox:latest                                                                                                                                         1.4s
 => => sha256:a01966dde7f8d5ba10b6d87e776c7c8fb5a5f6bfa678874bd28b33b1fc6dba34 828.50kB / 828.50kB                                                                                      0.2s
 => mkfile /foo                                                                                                                                                                         0.0s
 => mkfile /bar                                                                                                                                                                         0.0s
 => TestGroup                                                                                                                                                                           0.3s
 => => extracting sha256:a01966dde7f8d5ba10b6d87e776c7c8fb5a5f6bfa678874bd28b33b1fc6dba34                                                                                               0.0s
 => => merging                                                                                                                                                                          0.0s
 => => copying files 1.43MB                                                                                                                                                             0.0s
```

And here's the output when using `plain` mode:
```
#1 docker-image://docker.io/library/busybox:latest
#1 resolve docker.io/library/busybox:latest
#1 resolve docker.io/library/busybox:latest 1.4s done
#1 DONE 0.0s

#2 docker-image://docker.io/library/alpine:latest
#2 resolve docker.io/library/alpine:latest 1.4s done
#2 sha256:9b3977197b4f2147bdd31e1271f811319dcd5c2fc595f14e81f5351ab6275b99 0B / 2.72MB 0.2s
#2 sha256:9b3977197b4f2147bdd31e1271f811319dcd5c2fc595f14e81f5351ab6275b99 1.05MB / 2.72MB 0.3s
#2 sha256:9b3977197b4f2147bdd31e1271f811319dcd5c2fc595f14e81f5351ab6275b99 2.72MB / 2.72MB 0.5s done
#2 extracting sha256:9b3977197b4f2147bdd31e1271f811319dcd5c2fc595f14e81f5351ab6275b99 0.1s done
#2 DONE 0.5s

#3 mkfile /foo
#3 DONE 0.0s

#4 mkfile /bar
#4 DONE 0.0s

#5 TestGroup
#5 extracting sha256:a01966dde7f8d5ba10b6d87e776c7c8fb5a5f6bfa678874bd28b33b1fc6dba34 0.0s done
#5 merging done
#5 copying files 1.43MB 0.0s done
#5 DONE 0.3s

#1 docker-image://docker.io/library/busybox:latest
#1 sha256:a01966dde7f8d5ba10b6d87e776c7c8fb5a5f6bfa678874bd28b33b1fc6dba34 828.50kB / 828.50kB 0.2s done
#1 DONE 0.2s
```

tonistiigi commented:
I don't think we should push terms like "lazy" to the users. They shouldn't need to think about the internal implementation.

I would suggest 2 changes:
- Ability to mark a vertex as a "helper" of another vertex. In that case (in updated clients) the helper vertex will not show up but restart the main vertex. The time would be sum of all the `end-start` differences of the vertex group. We could use the same to group some of the `[internal]` blocks together for nicer UI. For the `copy+merge` case it would mean that only `COPY` row would be shown and when merge starts, its time is added to the `COPY`.
- Ensure that the order of items in the progressbar is related to how far the vertex is from the build result in llb graph. (After the sorting by whether the vertex is running has been done).

tonistiigi commented:
@sipsma Any update with this?

sipsma commented:
> @sipsma Any update with this?

I've been working on it intermittently but have been very busy with other stuff the last few weeks. Have more time this week so I will aim to get some updated code out for review by end of week (let me know if you have any time constraints for the v0.10 release I should know about).

In general, I'm structuring it in two commits; the first is the baseline behavior that at least results in merge+diff unlazies to show up under the right vertex (basically the same idea as how lazy blob refs work). After that I'll layer on some of the fancier progress grouping you were describing for merges specifically.

tonistiigi commented:
> the first is the baseline behavior that at least results in merge+diff unlazies to show up under the right vertex 

This is just the order of items, right? Because I don't think we need anything else special if we have grouping support.

sipsma commented:
> This is just the order of items, right? Because I don't think we need anything else special if we have grouping support.

For Merge yes. For Diff, I think grouping support won't cover all the cases (it would only help when diff is plugged into a merge, which while a majority of the cases is not all) and we will need the first commit.

tonistiigi commented:
> let me know if you have any time constraints for the v0.10 release I should know about

Targeting v0.10 RC for next week

sipsma commented:
Updated w/ the 2 commits described before; they don't overlap in terms of code so I'm happy to split them up into separate PRs if desired, but I wanted to develop them together because they do have indirect impacts on each other and I wanted to ensure the final output looked okay.

tonistiigi commented on solver/llbsolver/solver.go:
```diff
@@ -390,13 +395,14 @@ func oneOffProgress(ctx context.Context, id string) func(err error) error {
 }
 
-func inBuilderContext(ctx context.Context, b solver.Builder, name, id string, f func(ctx context.Context, g session.Group) error) error {
+func inBuilderContext(ctx context.Context, b solver.Builder, name, id string, pg *pb.ProgressGroup, f func(ctx context.Context, g session.Group) error) error {
```
What's the case where progressgroup is needed here? Isn't this function by definition for creating a separate progress row?

tonistiigi replied:
I don't understand. How is this different from merge with no progressgroup? ` e.Name()` is the exporter name in here and should always be shown and never combined with any other vertex.

sipsma replied:
I went back to disable this to show what the progress looks like without it and it looks totally fine without it now 😞 At one point unlazies that occurred during an export were not showing up in the progress group and I thought this was needed to fix it, but I must have fixed it some other way. I will confirm tomorrow morning and get rid of this if not needed.

sipsma replied:
Yeah, what happened was I fixed the export issue with this change originally, but then another fix I made at a later time inadvertently also served as a much simpler fix to it. So I updated with all this extra complication removed thankfully.

tonistiigi commented on solver/llbsolver/solver.go:
```diff
@@ -227,7 +228,11 @@ func (s *Solver) Solve(ctx context.Context, id string, sessionID string, req fro
-		if err := inBuilderContext(ctx, j, e.Name(), "", func(ctx context.Context, _ session.Group) error {
+		var pg *pb.ProgressGroup
+		if res.Ref != nil {
+			pg = res.Ref.ProgressGroup()
```
Not sure why you need to track progressgroup through the result when something like this is not needed for the name? To me, it should just be like a "secondary name" so implementation almost identical on daemon side.

sipsma replied:
This was the only way I could find to plumb the progress group metadata through to the `inBuilderContext` call below, which as mentioned in the other comment is needed so that progress groups show up correctly when an export forces an unlazy. I'm very open to suggestions on different ways of plumbing it.

> To me, it should just be like a "secondary name" so implementation almost identical on daemon side.

I'm not following, can you explain in more detail.

tonistiigi commented on util/progress/progressui/display.go:
```diff
@@ -222,7 +262,36 @@ func (t *trace) triggerVertexEvent(v *client.Vertex) {
+				group = &vertexGroup{
+					vertex: &vertex{
+						Vertex: &client.Vertex{
+							Digest:    digest.Digest(identity.NewID()),
```
why isn't this group id?

tonistiigi commented on util/progress/progressui/display.go:
```diff
@@ -248,6 +317,19 @@ func (t *trace) update(s *client.SolveStatus, termWidth int) {
+		group := t.groups[groupID]
+		newVtx := group.mergeSubVtxs()
+		t.triggerVertexEvent(newVtx)
+		if newVtx.Started != nil && group.Started == nil {
```
Don't understand this condition. Can't it run multiple times, resulting multiple `append()` calls.

It might have been easier to track all the `client.Vertex` that create a row with the same object. Adding support reference counting for the case where multiple start/complete need to be handled. Without the progress group there is just one reference.

tonistiigi replied:
> this is the same basic logic that was already in place for vertexes not in a group above.

I see a `prev` check also in the previous logic.

sipsma replied:
The `prev` check is the equivalent of the `group.Started` check here. But now that I understand what we're going for in terms of the start/stop interval timing this will need some updates, which I'm working on.

tonistiigi commented on util/progress/progressui/display.go:
```diff
@@ -159,6 +162,42 @@ func (v *vertex) update(c int) {
+			newVtx.Started = subVtx.Started
+		}
+
+		// Group is considered not completed if any vtxs are not completed. If all vtxs
```
The duration if the group is not `max(completed)-min(started)`. We need to sum the times when the group goes to completed and add the current time since it last went to started. We should do the same for non-groups as well when they go to started state multiple times. Possibly tracking both cases with the same object.

sipsma replied:
> We need to sum the times when the group goes to completed and add the current time since it last went to started.

I'm not following. Are you saying the total time for a vertex group should be the sum of the intervals each group member was running for? If so, I feel like that could be extremely confusing. Say there's 5 file-ops in a progress group that each take 2 seconds, but the file ops all run in parallel. The user would see that their progress group took 10 seconds but in reality it only took 2 seconds. This might make sense if they could see the sub vertexes in the group (and that their subtimes add to 10), but by design they can't see that.

sipsma replied:
Are you saying that we should sum the start/stop times but account for overlap? I.e. if the start stop intervals of the group members are
1. (0, 2)
1. (1, 2)
1. (3, 4)

Then the total time should be displayed as 3? If so, that makes more sense to me.

tonistiigi replied:
Not total of everything but it should only be active time.

Eg. common case: copy to scratch + merge on top of base. Copy to scratch will run instantly on the beginning of the build because it has no dependencies. Lets say it takes 0.5 seconds. Then base is built, maybe takes 5 min. Then merge can run with another 0.5s. The total time for the progressgroup should be 1s. Not 5:01.

When 2 refs are active at the same time then we do not need to (and shouldn't) double count. If vertex1 starts at 0:00 and ends in 0:10 and vertex2 in 0:05 and ends in 0:15 then total time is 15sec.

Same case is for the non-groups. If the vertex is started for cachemap and later for exec then progress should show the sum. This happens on image pull for example and is confusing atm.

tonistiigi replied:
Yes. Same as your last comment. I didn't see it before.

sipsma replied:
Awesome, makes sense. I am still working on this update as it's a bit more involved, I'll ping once I push it.

sipsma replied:
@tonistiigi updated with an extra commit implementing this (works with both normal vertexes and progress groups). I tested locally by adding temporary `time.Sleep` statements to various places to see the vertexes starting/stopping. It seems to work by my eye but it's not something easy to automatically test so let me know if anything seems off.

tonistiigi commented:
![image](https://user-images.githubusercontent.com/585223/152482941-b95a0b91-3f00-4deb-9b5d-dc70ce6e94ca.png)

Is what I get every time on the first build.

Otherwise, groups seem to work in https://github.com/moby/buildkit/pull/2596 but the issue with the out of order items remains and needs to be addressed separately

https://user-images.githubusercontent.com/585223/152484122-11e33a3b-65d0-49dd-a3ff-d674f7177484.mov

sipsma commented:
> Is what I get every time on the first build.

Just to confirm, the issue is that steps aren't appearing in the right order? so like in the screenshot/video `[internal] load build context` appears after `[1/5]` and `[2/5]` is after `[4/5]`, etc.? If so, I can definitely imagine how both the lazy progress and progress group handling could result in that.

tonistiigi commented:
The issue with the first screenshot (no progressgroup there) is that the `FROM` line is not complete although the build is done and the top line says `6/7`.

In the video (this uses mergeop) yes `3/5` and `4/5` are before `2/5` that is confusing to user as this is not how the order of lines is in the Dockerfile. Where the "build context" line is isn't that important I think..

sipsma commented:
> The issue with the first screenshot (no progressgroup there) is that the FROM line is not complete although the build is done and the top line says 6/7. Note that I only get this on the first build, if I have cache then this doesn't appear.

I can't reproduce it locally. I'm starting buildkit w/ `./hack/shell` using this PR's branch, and running `buildctl build --frontend dockerfile.v0 --local dockerfile=. --local context=.` with this Dockerfile:
```
FROM alpine
COPY Dockerfile /
COPY foo /
```

I ran 5 times and each time the output looks the same with everything being complete (blue):
```
[+] Building 2.0s (7/7) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                                    0.0s
 => => transferring dockerfile: 78B                                                                                                                                                     0.0s
 => [internal] load .dockerignore                                                                                                                                                       0.0s
 => => transferring context: 2B                                                                                                                                                         0.0s
 => [internal] load metadata for docker.io/library/alpine:latest                                                                                                                        1.5s
 => [1/3] FROM docker.io/library/alpine@sha256:21a3deaa0d32a8057914f36584b5288d2e5ecc984380bc0118285c70fa8c9300                                                                         0.5s
 => => resolve docker.io/library/alpine@sha256:21a3deaa0d32a8057914f36584b5288d2e5ecc984380bc0118285c70fa8c9300                                                                         0.0s
 => => sha256:9b3977197b4f2147bdd31e1271f811319dcd5c2fc595f14e81f5351ab6275b99 2.72MB / 2.72MB                                                                                          0.4s
 => => extracting sha256:9b3977197b4f2147bdd31e1271f811319dcd5c2fc595f14e81f5351ab6275b99                                                                                               0.1s
 => [internal] load build context                                                                                                                                                       0.0s
 => => transferring context: 102B                                                                                                                                                       0.0s
 => [2/3] COPY Dockerfile /                                                                                                                                                             0.0s
 => [3/3] COPY foo /                                                                                                                                                                    0.0s
```

Are you doing anything different? It could also possibly be a timing thing that just varies host to host. I wonder if the client progressui is being shut down before the final vertex `Completed` update is received.

> In the video (this uses mergeop) yes 3/5 and 4/5 are before 2/5 that is confusing to user as this is not how the order of lines is in the Dockerfile. Where the "build context" line is isn't that important I think..

I'll take a look at this, my initial suspicion is that it's a result of the code handling all the progress group updates after handling the non-group vertexes.

tonistiigi commented:
> I'll take a look at this, my initial suspicion is that it's a result of the code handling all the progress group updates after handling the non-group vertexes.

I don't think it is progress groups. It is expected that the copy to scratch runs early as it doesn't depend on other commands. But we should figure out how to avoid this and keep the order that makes sense in the build graph.

I'll see if I can debug the first case more on my side.

sipsma commented:
> I don't think it is progress groups. It is expected that the copy to scratch runs early as it doesn't depend on other commands. But we should figure out how to avoid this and keep the order that makes sense in the build graph.

Ah right. It might not be that hard to add another field to LLB metadata that specifies an order progress vertexes (and groups) should appear in. The progressui trace can then just sort its vertex list by that index.

sipsma commented:
I actually hit the first case where the `FROM` case never completes, it seems like it happens maybe 1/20 runs for me. I did some println debugging [between these two lines](https://github.com/moby/buildkit/blob/a640b47cb19c4f0ff47f2444f3215ee851598a8e/util/progress/progressui/display.go#L225-L226) to see each vertex received by the progress ui and found that when things do work the relevant vertexes received have start/complete times set to:
```
Started:<nil> Completed:<nil>

Started:2022-02-04 21:09:50.120528645 +0000 UTC Completed:<nil>

Started:2022-02-04 21:09:50.120528645 +0000 UTC Completed:2022-02-04 21:09:50.133164684 +0000 UTC 

Started:2022-02-04 21:09:50.133207933 +0000 UTC Completed:<nil>

Started:2022-02-04 21:09:50.133207933 +0000 UTC Completed:2022-02-04 21:09:50.133233391 +0000 UTC

Started:2022-02-04 21:09:50.133292389 +0000 UTC Completed:<nil>

Started:2022-02-04 21:09:50.133292389 +0000 UTC Completed:2022-02-04 21:09:50.142885665 +0000 UTC

Started:2022-02-04 21:09:50.145763182 +0000 UTC Completed:<nil>

Started:2022-02-04 21:09:50.145763182 +0000 UTC Completed:2022-02-04 21:09:50.509056664 +0000 UTC 

Started:2022-02-04 21:09:50.145763182 +0000 UTC Completed:<nil>

Started:2022-02-04 21:09:50.145763182 +0000 UTC Completed:2022-02-04 21:09:50.582821026 +0000 UTC
```

But when it doesn't complete they are:
```
Started:<nil> Completed:<nil>

Started:2022-02-04 21:09:58.775115015 +0000 UTC Completed:<nil>

Started:2022-02-04 21:09:58.789572264 +0000 UTC Completed:<nil>

Started:2022-02-04 21:09:58.789572264 +0000 UTC Completed:2022-02-04 21:09:58.799409704 +0000 UTC 

Started:2022-02-04 21:09:58.802618255 +0000 UTC Completed:<nil>

Started:2022-02-04 21:09:58.802618255 +0000 UTC Completed:2022-02-04 21:09:59.16156281 +0000 UTC 

Started:2022-02-04 21:09:58.802618255 +0000 UTC Completed:<nil>

Started:2022-02-04 21:09:58.802618255 +0000 UTC Completed:2022-02-04 21:09:59.232716669 +0000 UTC 
```

Not sure what's happening differently that causes the vertexes received to be different in the broken case.

But the reason this happens now is because the commit to support start/stop intervals in progress results in all intervals being merged and a complete time is never received for the start time of `21:09:58.775115015`. The vertex as a whole is therefore still considered to be running because that first interval never stopped.

It seems like figuring out why the vertexes received by the progress code are different in the broken case would be a good next step. Let me know if there's any known cases where that's expected to happen.

tonistiigi commented:
Is this all from same vertex? If it is then it already looks wrong that it switches so much. A single `FROM` should only go active twice.

If the server is already sending the wrong data then an additional backward-compatible way to fix it would be that if there is no progress-group then for a single vertex a completed time always marks it as finished.

sipsma commented:
> Is this all from same vertex? If it is then it already looks wrong that it switches so much. A single FROM should only go active twice.

Yeah those are all from just the vertex for the `FROM` statement, I filtered out the others. The only additional detail I gathered was that it seemed like the vertex with non-nil start that never got a corresponding complete always came from the `CacheMap` method [here](https://github.com/moby/buildkit/blob/a640b47cb19c4f0ff47f2444f3215ee851598a8e/solver/jobs.go#L760) (added some extra metadata to the vertex to figure this out). Haven't found anything there that would explain this more though. It might be worth looking at [this server-side code](https://github.com/moby/buildkit/blob/a640b47cb19c4f0ff47f2444f3215ee851598a8e/solver/progress.go#L15) to see if the vertexes make it there or if the problem is upstream from that point.

> If the server is already sending the wrong data then an additional backward-compatible way to fix it would be that if there is no progress-group then for a single vertex a completed time always marks it as finished.

Pushed an extra commit with this workaround for now.

tonistiigi commented:
I looked the server side. The reason it goes active for 5 times is:
- cachemap 2x, it is called 2 times for sources because sources have 2 keys
- exec
- 2x from unlazy handler https://github.com/moby/buildkit/blob/master/util/progress/controller/controller.go#L49

As usual, now I'm having trouble triggering the race so don't know what path produces the starting without matching completed time.

sipsma commented:
I may have figured out the underlying issue. I added logging to the `notifyStarted` and `notifyCompleted` funcs in `solver/jobs.go` and saw that when the missing completion case happens in progress, the functions are actually all running and the missing completion does get written [here](https://github.com/moby/buildkit/blob/dc32a321a00fb78058f1034dc0912ef0fd19fcc8/solver/jobs.go#L952).

I think the issue is on [this line](https://github.com/moby/buildkit/blob/dc32a321a00fb78058f1034dc0912ef0fd19fcc8/util/progress/progress.go#L248). `CacheMap` runs twice (as expected) but both instances of it use the same `ID` field for the progress struct, so a race is possible where the map value for the key of `ID` gets overwritten by the second cache map run before the progress written by the first cache map run gets picked up by the progress reader.

I made a change to just use a random ID in those `notifyStarted` and `notifyCompleted` funcs (which is hacky but works because [ID is not used in those cases](https://github.com/moby/buildkit/blob/dc32a321a00fb78058f1034dc0912ef0fd19fcc8/solver/progress.go#L36-L38)) and can no longer reproduce the issue after 50 attempts.

Just need to figure out a better solution than using a random ID now, will push this fix soon.

sipsma commented:
Pushed the fix (and removed the previous commit with the workaround in the progress ui). I updated to just include the start time (in unix nanos) with the ID, so then we can still dedupe multiple updates within the same start/stop interval, but won't dedupe across start/stop intervals.

tonistiigi commented on solver/jobs.go:
```diff
@@ -933,7 +934,10 @@ func notifyStarted(ctx context.Context, v *client.Vertex, cached bool) {
-	pw.Write(v.Digest.String(), *v)
+	// id dedupes progress updates; use of the start time ensures that distinct
+	// stop/start intervals of the vertex don't get deduped with one another.
+	id := v.Digest.String() + strconv.Itoa(int(v.Started.UnixNano()))
```
I'm not sure about using started time there. Might as well just use random then. If this is for the second cache key then what if we just pass a suffix for that case. Then at least the caller side still controls the ID. Could even be variadic I think.

tonistiigi commented on util/progress/progressui/display.go:
```diff
@@ -159,6 +167,160 @@ func (v *vertex) update(c int) {
+// interval's stop time is nil, it is treated as positive infinity and consumes
+// any intervals after it. Intervals with nil start times are ignored and not
+// returned.
+func mergeIntervals(intervals []interval) []interval {
```
Instead of these calculations wouldn't it be easier to just add the last `compled-started` diff every time vertex goes to completed state. If vertex is active then current time is `now-started+previous_sum`. Not a blocker but looks like it would be faster and simpler.

tonistiigi replied:
Do you think the extra complexity might become an issue for this for big builds? Does this get called repeatedly after every 100ms on tty mode?

sipsma replied:
It will get called for each `job` representing a vertex update (i.e. not status update, log, etc.) on each ui refresh. I just quickly ran a benchmark test using the `complex` test case in the unit tests I added and got 775 ns/op (in a linux VM running on an m1 macbook). That test case has 15 start/stop intervals to merge, whereas in the real world we'd usually expect at most 5. But round that up to 1 microsecond anyways, then assume that an extremely large build has 100 vertex `job`s per update and it still only takes 100 microseconds. Let me know if you think the estimate is off, but I'd be biased towards thinking the performance shouldn't make a difference here.

tonistiigi commented on util/progress/progressui/display.go:
```diff
@@ -103,28 +103,31 @@ type displayInfo struct {
 	nextIndex     int
 	updates       map[digest.Digest]struct{}
 	modeConsole   bool
+	groups        map[string]*vertexGroup // group id -> group
```
If all items would be "groups" and vertex without a progressgroup would just be a group with only one item I think some duplications could be avoided and maybe easier to understand. Not needed atm unless you think it is easy but I think worth considering as optimization.

We can assume that vertex digest and progress group id never collide.

sipsma commented:
> Are you working on this?

I haven't been but if you're good with that approach I'm happy to implement it. I should have time to send out a PR early next week.

tonistiigi commented:
@sipsma I'm not sure if there is a way to just do it automatically. We could compute how far the vertex is from the final build result that should be correct but that doesn't define how the multiple inputs of a mergeop should be sorted. Atm. the most critical need is something that would put `COPY --link` in Dockerfile to the correct place where it is in the file.

Note that there is some sorting fixes already in place iirc. It should be that when there is a lot of screen space it pushes the active items down so they are visible. That should still take precedence over the order defined in LLB. We need to avoid the case where the user doesn't see what is currently running as the whole screen is filled with already completed progress rows.

sipsma commented:
> Note that there is some sorting fixes already in place iirc. It should be that when there is a lot of screen space it pushes the active items down so they are visible. That should still take precedence over the order defined in LLB. We need to avoid the case where the user doesn't see what is currently running as the whole screen is filled with already completed progress rows.

Makes sense. Taking a look at this today.
